File: LV 2/[1차] 프렌즈4블록.py
from collections import deque

def check(board, m, n):
    dx = [0, 0, 1, 1]
    dy = [0, 1, 0, 1]
    ch = board[m][n]
    count = 0
    for i in range(4):
        x = m + dx[i]
        y = n + dy[i]
        if x >= 0 and x <= len(board)-1 and y >= 0 and y <= len(board[0])-1:
            if board[x][y] == ch:
                count += 1
    
    if count == 4:
        return True
    return False

def remove(board, m, n):
    dx = [0, 0, 1, 1]
    dy = [0, 1, 0, 1]
    for i in range(4):
        board[m+dx[i]][n+dy[i]] = '-'
    
def solution(m, n, board):
    board = [list(b) for b in board]
    answer = 0
    
    while True:
        currect = []
        # 2x2 검사
        for i in range(m):
            for j in range(n):
                if board[i][j] == '-':
                    continue
                if check(board, i, j):
                    currect.append([i, j])

        if len(currect) == 0:
            # -의 개수 구해줌
            for b in board: 
                answer += b.count('-')
            break
            
        # 검사결과 제거
        for cur in currect:
            remove(board, cur[0], cur[1])

        # - 붙여주는 과정
        columns = [[] for _ in range(n)]
        for j in range(n):
            for i in range(m):
                if board[i][j] != '-':
                    columns[j].append(board[i][j])
            if len(columns[j]) < m:
                columns[j] = ['-'] * (m - len(columns[j])) + columns[j] # 길이 '-' 붙여서 맞추어줌
        

        # - 제거한걸 원래 board에 적용
        for i in range(m):
            for j in range(n):
                board[i][j] = columns[j][i]
                
    return answer

File: LV 2/[1차] 프렌즈4블록.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

class Solution {

    private static final char BLANK = ' ';

    private boolean check(int m, int n, char[][] board) {
        char base = board[m][n];
        for (int i = m; i < m+2; i++) {
            for (int j = n; j < n+2; j++) {
                if (board[i][j] != base) {
                    return false;
                }
            }
        }
        return true;
    }

    private int destroy(int m, int n, char[][] board) {
        int destroyCount = 0;
        for (int i = m; i < m+2; i++) {
            for (int j = n; j < n+2; j++) {
                if (board[i][j] != BLANK) {
                    destroyCount++;
                }
                board[i][j] = BLANK;
            }
        }

        return destroyCount;
    }

    private void tuning(char[][] board) {
        for (int i = board.length-1; i >= 0; i--) {
            for (int j = 0; j < board.length; j++) {
                for (int k = i+1; k < board.length; k++) {
                    if (board[k][j] == BLANK) {
                        board[k][j] = board[k-1][j];
                        board[k-1][j] = BLANK;
                        continue;
                    }
                    break;
                }
            }
        }
    }

    public int solution(int m, int n, String[] board) {

        int answer = 0;

        List<char[]> temp = Arrays.stream(board)
                .map(String::toCharArray)
                .collect(Collectors.toList());
        char[][] convertBoard = temp.toArray(s -> new char[0][]);

        while (true) {
            List<int[]> candidates = getCandidate(convertBoard);

            if (candidates.size() == 0) {
                break;
            }

            for (int[] candidate : candidates) {
                answer += destroy(candidate[0], candidate[1], convertBoard);
            }
            
            tuning(convertBoard);
        }

        return answer;

    }

    private List<int[]> getCandidate(char[][] board) {
        List<int[]> candidates = new ArrayList<>();
        for (int i = 0; i < board.length - 1; i++) {
            for (int j = 0; j < board[0].length - 1; j++) {
                if (board[i][j] == BLANK) {
                    continue;
                }
                if (check(i, j, board)) {
                    candidates.add(new int[]{i, j});
                }
            }
        }
        return candidates;
    }
}
